File: fast_barcode_scanner/android/src/main/kotlin/com/jhoogstraat/fast_barcode_scanner/ImageHelper.kt
import android.content.Context
import android.graphics.ImageFormat
import android.graphics.Rect
import android.graphics.YuvImage
import android.media.Image
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class ImageHelper {
    private var savedCodes: MutableMap<String, String> = mutableMapOf()

    private object Holder {
        val INSTANCE = ImageHelper()
    }

    companion object {
        @JvmStatic
        fun getInstance(): ImageHelper {
            return Holder.INSTANCE
        }
    }

    // Store image to the path with barcode as filename
    private fun storeImage(imageBytes: ByteArray?, key: String, context: Context) {
        val externalFilesDirectory = context.cacheDir;
        val imageFile: File

        try {
            externalFilesDirectory.mkdirs()
            val barcodeDirectory = File(externalFilesDirectory, "barcode_images")
            if (!barcodeDirectory.exists()) {
                barcodeDirectory.mkdirs()
            }

            imageFile = File.createTempFile(key, ".jpeg", barcodeDirectory)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

        try {
            FileOutputStream(imageFile).use { fos ->
                fos.write(imageBytes)
            }
            savedCodes[key] = imageFile.absolutePath
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    private fun isImageSaved(code: String): Boolean {
        return savedCodes.contains(code)
    }

    // Retrieve image from cache by barcode
    fun retrieveImagePath(code: String): String? {
        return savedCodes[code]
    }

    fun clearCache(context: Context) {
        savedCodes.clear()
        val externalFilesDirectory = context.cacheDir
        val barcodeDirectory = File(externalFilesDirectory, "barcode_images")
        if (barcodeDirectory.exists()) {
            barcodeDirectory.deleteRecursively()
        }
    }

    suspend fun storeImageToCache(image: Image, code: String, context: Context) {
        if (isImageSaved(code)) {
            return
        }
        if (image.format == ImageFormat.YUV_420_888) {
            withContext(Dispatchers.IO) {
                image.use { img ->
                    val yBuffer = img.planes[0].buffer // Y
                    val uBuffer = img.planes[1].buffer // U
                    val vBuffer = img.planes[2].buffer // V

                    val ySize = yBuffer.remaining()
                    val uSize = uBuffer.remaining()
                    val vSize = vBuffer.remaining()

                    val nv21 = ByteArray(ySize + uSize + vSize)

                    // Copy Y channel
                    yBuffer[nv21, 0, ySize]

                    // Copy VU channel (assuming NV21 format)
                    vBuffer[nv21, ySize, vSize]
                    uBuffer[nv21, ySize + vSize, uSize]

                    val yuvImage = YuvImage(nv21, ImageFormat.NV21, img.width, img.height, null)
                    val out = ByteArrayOutputStream()
                    yuvImage.compressToJpeg(Rect(0, 0, img.width, img.height), 100, out)
                    val jpegBytes = out.toByteArray()
                    storeImage(jpegBytes, code, context)
                }
            }
        } else {
            throw IllegalArgumentException("Unsupported image format")
        }
    }
}

File: fast_barcode_scanner/ios/Classes/Utilities/ImageHelper.swift
import UIKit

class ImageHelper {
    private var savedCodes: [String: String] = [:]

    static let shared = ImageHelper()

    private init() {}

    // Store image to the path with barcode as filename
    private func storeImage(imageBytes: Data?, key: String) {

        let fileManager = FileManager.default
        let documentDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first

        do {
            let barcodeDirectory = documentDirectory.appendingPathComponent("barcode_images")
            try fileManager.createDirectory(at: barcodeDirectory, withIntermediateDirectories: true, attributes: nil)

            let imageFile = barcodeDirectory.appendingPathComponent(key + ".jpeg")
            try imageBytes.write(to: imageFile)

            savedCodes[key] = imageFile.absoluteString
        } catch {
            print("Error storing image: \(error)")
        }
    }

    private func isImageSaved(code: String) -> Bool {
        return savedCodes.keys.contains(code)
    }

    // Retrieve image from cache by barcode
    func retrieveImagePath(code: String) -> String? {
        return savedCodes[code]
    }

    func clearCache() {
        savedCodes.removeAll()

        let fileManager = FileManager.default
        let documentDirectory = fileManager.urls(for: .documentDirectory, in: .userDomainMask).first
        let barcodeDirectory = documentDirectory.appendingPathComponent("barcode_images")

        do {
            try fileManager.removeItem(at: barcodeDirectory)
        } catch {
            print("Error clearing cache: \(error)")
        }
    }

    func storeImageToCache(image: UIImage, code: String) {
        if isImageSaved(code) {
            return
        }

        // Convert UIImage to JPEG Data
        guard let jpegBytes = image.jpegData(compressionQuality: 1.0) else {
            print("Error converting image to JPEG")
            return
        }

        storeImage(imageBytes: jpegBytes, key: code)
    }
}
